RemoteDisplayListRecorder sends trivial transforms to GPUP

kkinnunen-apple opened the pull request:
#### 0d267f7455a4b8f173a0151f7e39a8e651691e3f
<pre>
RemoteDisplayListRecorder sends trivial transforms to GPUP
<a href="https://bugs.webkit.org/show_bug.cgi?id=256362">https://bugs.webkit.org/show_bug.cgi?id=256362</a>
rdar://108945948

Reviewed by Simon Fraser and Said Abou-Hallawa.

Avoid sending trivial translates, scales, rotates.
Avoids small amount of CPU work.

* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::translate):
(WebCore::DisplayList::Recorder::rotate):
(WebKit::RemoteLayerBackingStore::drawInContext):
* Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp:

Canonical link: <a href="https://commits.webkit.org/263796@main">https://commits.webkit.org/263796@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9dc50aeb39d36e8c0d894a79e663ade4ae543cec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5781 "29 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5946 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6134 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7340 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6173 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5781 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6168 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5913 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/7955 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5886 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5934 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5225 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7394 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3411 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5208 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13116 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5279 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5285 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/7459 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5732 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5175 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1366 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9294 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5536 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -300,18 +300,24 @@ void Recorder::restore()
 
 void Recorder::scale(const FloatSize& size)
 {
+    if (size == FloatSize { 1.f, 1.f })
```
Should we use areEssentiallyEqual() instead?

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -300,18 +300,24 @@ void Recorder::restore()
 
 void Recorder::rotate(float angleInRadians)
 {
+    if (!angleInRadians)
```
Should we check if angleInRadians is multiple of 2pi?

smfr replied:
Some of these transforms may have a Save/Restore associated with them. I wonder if we can optimize those away too.

kkinnunen-apple replied:
> Should we check if angleInRadians is multiple of 2pi?

I can add those but they're a bit more risky..

> Some of these transforms may have a Save/Restore associated with them. I wonder if we can optimize those away too.

I haven't seen meaningful cases of exactly these.
I'm working on a related case,` save, clip, restore, -> remove all`, which will take care of the above. However, that needs quite a bit more machinery than will fit in this patch.

kkinnunen-apple committed:
RemoteDisplayListRecorder sends trivial transforms to GPUP
https://bugs.webkit.org/show_bug.cgi?id=256362
rdar://108945948

Reviewed by Simon Fraser and Said Abou-Hallawa.

Avoid sending trivial translates, scales, rotates.
Avoids small amount of CPU work.

* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::translate):
(WebCore::DisplayList::Recorder::rotate):
(WebKit::RemoteLayerBackingStore::drawInContext):
* Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp:

Canonical link: https://commits.webkit.org/263796@main
